### PR TITLE
8264707: HotSpot Style Guide should permit use of lambda

### PR DESCRIPTION
**Pull request by kimbarrett**
Please review this proposal to permit the use of lambda expressions in
HotSpot code, with some restrictions and suggestions for good usage within
HotSpot code.  Lambda expressions were added in C++11, and provide a more
expressive syntax for local functions, with a number of use-cases where they
can improve readability by eliminating a lot of uninteresting boilerplate.

Some example uses are included, but are not part of the proposed change.
They will be removed from the PR before it is pushed.  (In particular, the
ScopeGuard utility uses move semantics, the use of which hasn't been
approved or even discussed.)  They are given to show some of the benefits
that might accrue from permitting the use of lambdas. In particular, they
highlight some of the code reduction that is possible. Some of these code
changes might be proposed in the future, using the normal PR process.

This is a modification of the Style Guide, so rough consensus among the
HotSpot Group members is required to make this change.  Only Group members
should vote for approval (via the github PR), though reasoned objections or
comments from anyone will be considered.  A decision on this proposal will
not be made before Wednesday 1-Sep-2021 at 12h00 UTC.

Since we're piggybacking on github PRs here, please use the PR review
process to approve (click on Review Changes > Approve), rather than sending
a "vote: yes" email reply that would be normal for a CFV.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8264707](https://bugs.openjdk.java.net/browse/JDK-8264707): HotSpot Style Guide should permit use of lambda

### Reviewers
 * [Stefan Karlsson](https://openjdk.java.net/census#stefank) (@stefank - **Reviewer**) ⚠️ Review applies to cc08f8b435eb0862d07a6ec6fc0af70b60c3b70e
 * [David Holmes](https://openjdk.java.net/census#dholmes) (@dholmes-ora - **Reviewer**) ⚠️ Review applies to cc08f8b435eb0862d07a6ec6fc0af70b60c3b70e
 * [Coleen Phillimore](https://openjdk.java.net/census#coleenp) (@coleenp - **Reviewer**) ⚠️ Review applies to 1fd7efbca4182d51f374ea271c2128a9e23faa1c
 * [Ioi Lam](https://openjdk.java.net/census#iklam) (@iklam - **Reviewer**) ⚠️ Review applies to 1fd7efbca4182d51f374ea271c2128a9e23faa1c
 * [Stefan Johansson](https://openjdk.java.net/census#sjohanss) (@kstefanj - **Reviewer**) ⚠️ Review applies to 1fd7efbca4182d51f374ea271c2128a9e23faa1c
 * [Erik Österlund](https://openjdk.java.net/census#eosterlund) (@fisk - **Reviewer**) ⚠️ Review applies to 1fd7efbca4182d51f374ea271c2128a9e23faa1c
 * [Igor Veresov](https://openjdk.java.net/census#iveresov) (@veresov - **Reviewer**) ⚠️ Review applies to 1fd7efbca4182d51f374ea271c2128a9e23faa1c
 * [Vladimir Kozlov](https://openjdk.java.net/census#kvn) (@vnkozlov - **Reviewer**) ⚠️ Review applies to 1fd7efbca4182d51f374ea271c2128a9e23faa1c

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/5144/head:pull/5144` \
`$ git checkout pull/5144`

Update a local copy of the PR: \
`$ git checkout pull/5144` \
`$ git pull https://git.openjdk.java.net/jdk pull/5144/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 5144`

View PR using the GUI difftool: \
`$ git pr show -t 5144`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/5144.diff">https://git.openjdk.java.net/jdk/pull/5144.diff</a>

</details>

### CODE REVIEW

**Review by stefank — APPROVED** [review]

File: doc/hotspot-style.md [hunk 1]
```diff
@@ -703,6 +701,274 @@ Some relevant sections from cppreference.com:
...
+* Generic lambda expressions are permitted.
+* Lambda expressions should be relatively simple.
+* Anonymous lambda expressions should not overly clutter the enclosing expression.
+* An anonymous lambda expression must not be directly invoked.
+* Bind expressions are forbidden.
+
+Single-use function objects can be defined locally within a function.  This
+is an alternative to having a function or function object class defined at
+class or namespace scope.  Instead, a function class definition is placed
+directly at the point of use.
```
Comment (dholmes-ora) [inline doc/hotspot-style.md:721]: Suggested rewording:

> ... namespace scope, instead placing it directly at the point of use.

**Review by dholmes-ora — APPROVED** [review]

File: doc/hotspot-style.md (hunk 1 above)
Comment (kimbarrett) [reply doc/hotspot-style.md:721]: The wording was intentional; "it" would refer to "a function or function object class", but a *function* can't be locally defined.  But it should be consistent about "function object class" vs "function class".  The former is the terminology used by the Standard.

Comment (dholmes-ora) [reply doc/hotspot-style.md:721]: There should be a consistent subject for each of those sentences, such that "it" would unambiguously be that subject, and then you can combine the second and third sentences into one.

Comment (kimbarrett) [reply doc/hotspot-style.md:721]: Changing the paragraph to this is probably better.  The third sentence didn't really add anything.

"Single-use function objects can be defined locally within a function, directly at the point of use.  This is an alternative to having a function or function object class defined at class or namespace scope."

File: doc/hotspot-style.md [hunk 2]
```diff
@@ -703,6 +701,274 @@ Some relevant sections from cppreference.com:
 Although related, the use of `std::initializer_list` remains forbidden, as
 part of the avoidance of the C++ Standard Library in HotSpot code.
 
+### Local Function Objects
+
+* Local function objects, including lambda expressions, may be used.
+* Lambda expressions must only be used as a downward value.
+* Prefer `[&]` as the capture list of a lambda expression.
```
Comment (navyxliu) [inline doc/hotspot-style.md:708]: hi, @kimbarrett , 
Why does it sound like a suggestion instead of a mandate? 
You explain why we should use [&] and all your examples are implicit capture reference.

Comment (kimbarrett) [reply doc/hotspot-style.md:708]: That's intentional. Capture lists were by far the single most contentious part
of the Oracle-internal discussions about lambda usage. Making this a
suggestion rather than a mandate, along with the rationale, leaves open the
possibility of using other forms *with good reason*, and preferably after a
period of folks getting used to the "usual" form. There are HotSpot developers
who aren't familiar with lambda-like constructs, or are not comfortable with
the introduction of yet more non-alphanumeric syntax that resembles line
noise; limiting the capture list features being used helps them. On the other
hand, there have been proponents of nearly every capture list feature for some
use case or another.

**Comment by theRealAph** [comment]
So here's an oink in the flightment:

```
macroAssembler_aarch64_aes.o: Error: Use of global operators new and delete is not allowed in Hotspot:
                 U operator delete(void*)
                 U operator new(unsigned long)
See: /Users/aph/theRealAph-jdk/make/hotspot/lib/CompileJvm.gmk

```

This happened on MacOS/AArch64, and was caused by an apparently innocuous Lambda. GCC doesn't generate new and delete for this construct, but AArch64 clang does for some reason. And I guess it's true that C++ compilers are free to do this, and even if one compiler doesn't do so today, it might tomorrow.

What should we do? At least for my application, it doesn't matter if new and delete are used, but in some cases it might. Do we need a blanket prohibition against new and delete, when the programmer has no control over it?

**Comment by kimbarrett** [comment]
> So here's an oink in the flightment:
> 
> ```
> macroAssembler_aarch64_aes.o: Error: Use of global operators new and delete is not allowed in Hotspot:
>                  U operator delete(void*)
>                  U operator new(unsigned long)
> See: /Users/aph/theRealAph-jdk/make/hotspot/lib/CompileJvm.gmk
> ```
> 
> This happened on MacOS/AArch64, and was caused by an apparently innocuous Lambda. GCC doesn't generate new and delete for this construct, but AArch64 clang does for some reason. And I guess it's true that C++ compilers are free to do this, and even if one compiler doesn't do so today, it might tomorrow.
> 
> What should we do? At least for my application, it doesn't matter if new and delete are used, but in some cases it might. Do we need a blanket prohibition against new and delete, when the programmer has no control over it?

Bleh!  Why would a lambda need new/delete?

Can you give more details about how you encountered this?  I've not seen
anything of the sort.

The purpose of that error is to prevent HotSpot code from bypassing NMT. NMT
can't just override the global operator new/delete because an application
might want to use its own for its native code.

File: doc/hotspot-style.md (hunk 2 above)
Comment (navyxliu) [reply doc/hotspot-style.md:708]: Got it. Thanks for the explanation.

File: src/hotspot/share/compiler/compilerOracle.cpp [hunk 3]
```diff
@@ -778,25 +779,13 @@ void CompilerOracle::print_parse_error(char* error_msg, char* original_line) {
...
-    }
-};
-
 void CompilerOracle::parse_from_line(char* line) {
   if (line[0] == '\0') return;
   if (line[0] == '#')  return;
 
-  LineCopy original(line);
+  char* original_line = os::strdup(line, mtInternal);
+  auto g = make_guard([&] { os::free(original_line); });
```
Comment (coleenp) [inline src/hotspot/share/compiler/compilerOracle.cpp:787]: I have to admit that this usage is very mysterious.  I know you're going to take it out but using lambdas for something that's really hard to parse doesn't seem like a win.  Does save lines of code though.

File: src/hotspot/share/gc/g1/g1SATBMarkQueueSet.cpp [hunk 4]
```diff
@@ -97,24 +97,9 @@ static inline bool requires_marking(const void* entry, G1CollectedHeap* g1h) {
...
-    return discard_entry(entry, _g1h);
-  }
-};
-
 void G1SATBMarkQueueSet::filter(SATBMarkQueue& queue) {
-  apply_filter(G1SATBMarkQueueFilterFn(), queue);
+  auto requires_discard = [&] (const void* entry) {
+    return !requires_marking(entry, _g1h) || _g1h->is_marked_next(cast_to_oop(entry));
+  };
+  apply_filter(requires_discard, queue);
```
Comment (coleenp) [inline src/hotspot/share/gc/g1/g1SATBMarkQueueSet.cpp:104]: This on the other hand, is a nice usage and saves a lot of boilerplate, and it's not like operator() was that easy to read in the first place.

**Review by coleenp — APPROVED** [review]
I approve of adding lambdas as specified in the coding standard.

File: src/hotspot/share/compiler/compilerOracle.cpp (hunk 3 above)
Comment (kimbarrett) [reply src/hotspot/share/compiler/compilerOracle.cpp:787]: I think a different approach may be possible, that would have usage like this:
```
char* original_line = os::strdup(line, mtInternal); 
ScopeGuard guard([&] { os::free(original_line); });
```
This would use something like (or maybe directly use) a solution to the type-erased lambda capturing problem a la std::function.

**Comment by rose00** [comment]
Reviewed. Thank you for collecting so much practical and useful advice.

**Review by iklam — APPROVED** [review]

**Review by kstefanj — APPROVED** [review]
Vote: yes

**Review by fisk — APPROVED** [review]

**Review by veresov — APPROVED** [review]

**Review by vnkozlov — APPROVED** [review]
Approved.

**Comment by kimbarrett** [comment]
Thanks @vnkozlov and all the other reviewers and commenters.

As promised, I'll be backing out the example code changes before integrating.  I'll also regenerate the html version of the style guide and include that in the integration.

File: src/hotspot/share/compiler/compilerOracle.cpp (hunk 3 above)
Comment (coleenp) [reply src/hotspot/share/compiler/compilerOracle.cpp]: Yes, this second usage looks like the typical RAII object that is familiar in the code.
